sklearn/feature_selection/_base.py

```diff
@@ -214,10 +223,24 @@ def _get_feature_importances(estimator, getter, transform_func=None, norm_order=
                 )
         else:
             getter = attrgetter(getter)
-    elif not callable(getter):
-        raise ValueError("`importance_getter` has to be a string or `callable`")
+        importances = getter(estimator)
 
-    importances = getter(estimator)
+    elif callable(getter):
+        # TODO(1.4): Replace the if-else below with
+        # importances = getter(estimator, X, y)
+        if len(signature(getter).parameters) == 3:
+            importances = getter(estimator, X, y)
+        else:
+            importances = getter(estimator)
```

Comment:
Unfortunately, `_get_feature_importances` is shared with `SelectFromModel`, which  only uses a one parameter. `SelectFromModel` can accept a prefitted estimator and use that for feature selection. In those cases, `SelectFromModel` never saw that data which means it can not use the three parameter callable.

It feels like a inconsistent API for `SelectFromModel` to only accept a callable with 1 parameter and `RFE` only accept a callable with 3 parameter. Also, there is some added complexity to `_get_feature_importances` if we want to restrict the number of parameters.

With that in mind, I am leaning toward your original idea of supporting both 1 and 3 parameters in `RFE`.

Reply:
Sounds good to me.

I have reverted the commits where I deprecated accepting a callable with 1 parameter.

What do you think about extending the documentation for the argument `importance_getter` as it is in this PR to `SelectFromModel` as well? Another option might be to deprecate accepting a callable with 1 parameter there as well.

Reply:
As noted in https://github.com/scikit-learn/scikit-learn/pull/21935/files/5c81eef0390b0b96b6e0ca74256430d0a9e5a88a..4e74568e8912082647a302f548d541c821718c99#r940622865, the way `SelectFromModel` can accept prefitted estimators means that it will always need to support a callable with one parameter. Specifically, if `SelectFromModel` is configured with a prefited model, then `SelectFromModel.transform` works without calling `fit` and seeing the training data.

For this PR, I prefer not to expand scope to `SelectFromModel`. Usually expanding scope makes a PR harder to merge.

Reply:
Makes sense. I that case I think the PR should be ready to be merged.

I also thought about adding an example about using a custom importance getter in the recursive feature elimination section, but I will open another PR about it once this one has been merged.

Reply:
@thomasjpfan 
Would it make sense to potentially wrap the callable `importance_getter` passed to `SelectFromModel` internally so that it can match the 3-argument signature forced by the internal API? The user experience wouldn't change and it would allow for a consistent internal API.

If we wanted to (in a future PR) we could even explore the path of allowing 3-argument callables for `SelectFromModel` when working with yet-unfit estimators.